libopenage/unit/producer.cpp

```diff
@@ -623,8 +623,29 @@ TerrainObject *BuildingProducer::place(Unit *u, std::shared_ptr<Terrain> terrain
 			if (!tc) {
 				return false;
 			}
+
+			// check if it's in "unbuildable terrain" blacklist
+			// TODO: match this with valid enum
+			if (tc->terrain_id == 1 || // water
+				tc->terrain_id == 4 || // shallows
+				tc->terrain_id == 14 || // deep water
+				tc->terrain_id == 15 || // medium water
+				tc->terrain_id == 18 // ice
+					) return false;
```

Comment:
This terrain blacklist is called ["terrain restriction"](https://github.com/SFTtech/openage/blob/fe0a66dfa9f4850e1002998245d2d82b7f4d9728/openage/convert/gamedata/terrain.py#L39). We don't use it yet, but maybe you can figure out how it works.

Also, would be nice for consistency reasons if you indented the `tc->terrain_id == 4` by spaces starting at the indent level of `if` to line up with `tc->terrain_id == 1`, see [here](https://github.com/SFTtech/openage/blob/master/doc/code_style/tabs_n_spaces.md).